### PR TITLE
Raspawar/add vlm support

### PR DESCRIPTION
**Pull request by raspawar**
## Support for vision-language models, those that can accept images and text as input and produce text.

these are akin to https://platform.openai.com/docs/guides/vision with three notable differences -

- images can be passed with img tags in the regular text content
- images can be passed as NVCF asset ids
not all model endpoints support all features, e.g. server-side download of images not available with 
adept/fuyu-8b, google/deplot, microsoft/kosmos-2, google/paligemma; some models endpoints restrict image size; some models support one and only image; some models do not support gif or webp; kosmos-2 does not support streaming

cc: @sumitkbh @mattf @dglogo

### CODE REVIEW

File: llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/llama_index/multi_modal_llms/nvidia/base.py [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+        description="The timeout, in seconds, for API requests.",
+        ge=0,
+    )
+    api_key: str = Field(default=None, description="The NVIDIA API key.", exclude=True)
+    system_prompt: str = Field(default="", description="System Prompt.")
+    api_base: str = Field(default=None, description="The base URL for NVIDIA API.")
+    additional_kwargs: Dict[str, Any] = Field(
+        default_factory=dict, description="Additional kwargs for the NVIDIA API."
+    )
+    default_headers: Optional[Dict[str, str]] = Field(
```
Comment (mattf) [inline llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/llama_index/multi_modal_llms/nvidia/base.py:142]: what's the use case for `default_headers`?

File: llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/llama_index/multi_modal_llms/nvidia/base.py [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+        default=None, description="The default headers for API requests."
+    )
+
+    _messages_to_prompt: Callable = PrivateAttr()
+    _completion_to_prompt: Callable = PrivateAttr()
```
Comment (mattf) [inline llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/llama_index/multi_modal_llms/nvidia/base.py:147]: how and where is `_completion_to_prompt` used?

File: llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/llama_index/multi_modal_llms/nvidia/base.py [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+    def stream_complete(
+        self, prompt: str, image_documents: Sequence[ImageNode], **kwargs: Any
+    ) -> CompletionResponseGen:
+        return self._stream_complete(prompt, image_documents, **kwargs)
+
+    def chat(
+        self,
+        **kwargs: Any,
+    ) -> Any:
+        raise NotImplementedError("This function is not yet implemented.")
```
Comment (mattf) [inline llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/llama_index/multi_modal_llms/nvidia/base.py:350]: please implement.

suggestion: rename _complete -> _chat, use generate_nvidia_multi_modal_chat_message + _chat in complete

File: llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/llama_index/multi_modal_llms/nvidia/base.py [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+
+    def stream_chat(
+        self,
+        **kwargs: Any,
+    ) -> Any:
+        raise NotImplementedError("This function is not yet implemented.")
```
Comment (mattf) [inline llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/llama_index/multi_modal_llms/nvidia/base.py:356]: please implement.

suggestion: rename _stream_complete -> _stream_chat, use generate_nvidia_multi_modal_chat_message + _stream_chat in stream_complete

File: llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/llama_index/multi_modal_llms/nvidia/base.py [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+
+    # ===== Async Endpoints =====
```
Comment (mattf) [inline llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/llama_index/multi_modal_llms/nvidia/base.py:358]: implement or open issue to finish implementation

**Review by mattf — CHANGES_REQUESTED** [review]
please address comments.

failing tests -
```
FAILED tests/test_multi_modal_nvidia.py::test_vlm_asset_id[invoke-content0-microsoft/phi-3-vision-128k-instruct] - TypeError: sequence item 0: expected str instance, function found
FAILED tests/test_multi_modal_nvidia.py::test_vlm_asset_id[stream-content0-microsoft/phi-3-vision-128k-instruct] - TypeError: sequence item 0: expected str instance, function found
```

File: llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/llama_index/multi_modal_llms/nvidia/base.py (hunk 1 above)
Comment (mattf) [reply llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/llama_index/multi_modal_llms/nvidia/base.py:148]: why give users control over headers used internally?

File: llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/llama_index/multi_modal_llms/nvidia/base.py [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+            api_base=base_url,
+            callback_manager=callback_manager,
+            default_headers=default_headers,
+            **kwargs,
+        )
+
+    @property
+    def _client(self) -> NVIDIAClient:
+        _client = NVIDIAClient(**self._get_credential_kwargs())
+        _client._custom_headers = {"User-Agent": "llama-index-multi-modal-llms-nvidia"}
```
Comment (mattf) [inline llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/llama_index/multi_modal_llms/nvidia/base.py:190]: it appears `_custom_headers` is not used. please move the user agent into NVIDIAClient's _get_headers.

**Review by mattf — CHANGES_REQUESTED** [review]

File: llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/tests/data/nvidia-picasso-large.png
[the remark this answers is not in the document]
Comment (raspawar) [reply llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/tests/data/nvidia-picasso-large.png:1]: got it will do

**Comment by raspawar** [comment]
@logan-markewich can u ptal. I donno why the coverage is considering the test cases also.

File: llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/README.md
[the remark this answers is not in the document]
Comment (raspawar) [reply llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/README.md:1]: Done!

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**